transports/janus_http.c

```diff
@@ -1186,15 +1209,18 @@ int janus_http_handler(void *cls, struct MHD_Connection *connection, const char
 	/* Suspend the connection and pass the ball to the core */
 	JANUS_LOG(LOG_HUGE, "Forwarding request to the core (%p)\n", msg);
 	gateway->incoming_request(&janus_http_transport, msg, msg, FALSE, root, &error);
-	gint64 waited = 0;
-	while(!msg->got_response) {
-		/* Still waiting for a response */
-		g_usleep(50000);
-		waited += 50000;
-		if(waited >= 10*G_USEC_PER_SEC) {
-			/* Don't wait forever */
+	/* Wait for a response (but not forever) */
+	struct timeval now;
+	gettimeofday(&now, NULL);
+	struct timespec wakeup;
+	wakeup.tv_sec = now.tv_sec+10;	/* Wait at max 10 seconds for a response */
+	wakeup.tv_nsec = now.tv_usec*1000UL;
+	while(TRUE) {
+		pthread_mutex_lock(&msg->wait_mutex);
+		int res = pthread_cond_timedwait(&msg->wait_cond, &msg->wait_mutex, &wakeup);
+		pthread_mutex_unlock(&msg->wait_mutex);
+		if(msg->got_response || res == ETIMEDOUT)
```

Comment:
Where you call `janus_condition_signal()` you should set `msg->got_response` while you hold the mutex, and here you should check `msg->got_response` also while holding the mutex, *before* waiting on the condition variable.

Otherwise, it's theoretically possible that the condition is signaled just before you take the mutex and wait on it.